File: python/tests/test_investments_data.py
import pytest

from rts_gmlc_case.investments import data


def test_source_dir_exists_and_has_all_required_files():
    src = data.investments_source_dir()
    for name in data.REQUIRED_FILES:
        assert (src / name).is_file(), name


def test_source_dir_matches_exactly_the_16_required_files():
    src = data.investments_source_dir()
    on_disk = {
        str(path.relative_to(src)) for path in src.rglob("*") if path.is_file()
    }
    assert on_disk == set(data.REQUIRED_FILES)


def test_missing_directory_raises_without_downloading(tmp_path, monkeypatch):
    monkeypatch.setattr(data, "REPO_ROOT", tmp_path)

    with pytest.raises(RuntimeError, match="does not exist"):
        data.investments_source_dir()


def test_partial_directory_raises_with_missing_names(tmp_path, monkeypatch):
    monkeypatch.setattr(data, "REPO_ROOT", tmp_path)
    root = tmp_path / "data" / "RTS-investments"
    (root / "capacitydata").mkdir(parents=True)
    (root / "hierarchy_rts.csv").write_text("")
    (root / "scalars.csv").write_text("")

    with pytest.raises(RuntimeError) as exc_info:
        data.investments_source_dir()

    message = str(exc_info.value)
    present = {"hierarchy_rts.csv", "scalars.csv"}
    for name in data.REQUIRED_FILES:
        if name in present:
            continue
        assert name in message

File: julia/RTSGMLCCase/test/test_investments_data.jl
@testset "investments source data" begin
    src = RTSGMLCCase.investments_source_dir()
    for name in RTSGMLCCase.REQUIRED_INVESTMENTS_FILES
        @test isfile(joinpath(src, name))
    end
end

@testset "investments source dir matches exactly the 16 required files" begin
    src = RTSGMLCCase.investments_source_dir()
    on_disk = Set{String}()
    for (root, _, files) in walkdir(src)
        for file in files
            push!(on_disk, relpath(joinpath(root, file), src))
        end
    end
    @test on_disk == Set(RTSGMLCCase.REQUIRED_INVESTMENTS_FILES)
end

@testset "missing investments directory raises without downloading" begin
    mktempdir() do repo_root
        err = nothing
        try
            RTSGMLCCase.investments_source_dir(repo_root)
        catch caught
            err = caught
        end
        @test err !== nothing
        message = sprint(showerror, err)
        @test occursin("does not exist", message)
    end
end

@testset "partial investments directory raises with missing names" begin
    mktempdir() do repo_root
        root = joinpath(repo_root, "data", "RTS-investments")
        mkpath(joinpath(root, "capacitydata"))
        write(joinpath(root, "hierarchy_rts.csv"), "")
        write(joinpath(root, "scalars.csv"), "")

        err = nothing
        try
            RTSGMLCCase.investments_source_dir(repo_root)
        catch caught
            err = caught
        end
        @test err !== nothing
        message = sprint(showerror, err)
        present = Set(["hierarchy_rts.csv", "scalars.csv"])
        for name in RTSGMLCCase.REQUIRED_INVESTMENTS_FILES
            name in present && continue
            @test occursin(name, message)
        end
    end
end
